CNV-25648: boot source doc improvements

ousleyp opened the pull request:
<!--- PR title format: [GH#<gh-issue-id>][BZ#<bz-issue-id>][OCPBUGS#<jira-issue-id>][OSDOCS#<jira-issue-id>]: <short-description-of-the-pr> --->

<!--- If your changes apply to the latest release and/or in-development version of OpenShift, open your PR against the `main` branch.

* For more details about the information requested in this template, see:
  https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/create_or_edit_content.adoc#submit-PR --->

Version(s): 4.13+ (minimum viable cherrypick + potentially 4.12 and/or 4.11)
<!--- Specify the version or versions of OpenShift your PR applies to. -->

Issue: [CNV-25648](https://issues.redhat.com//browse/CNV-25648)
<!--- Add a link to the Bugzilla, Jira, or GitHub issue, if applicable. --->

Link to docs preview: 
- Autoupdate assembly: https://ousleyp.github.io/previews/cnv-25648/virt/vm_templates/virt-automatic-bootsource-updates.html
<!--- Add direct link(s) to the exact page(s) with updated content from the preview build. --->

QE review: 
- [x] QE has approved this change.
<!--- QE approval is required to merge a PR except for changes that do not impact the meaning of the docs. --->

Additional information: 
- I am splitting this Jira into two or three stories. This PR will only cover this reorg/rework of the auto-update docs.
- Note for peer reviewer: This is an improvement story where most of the content already existed; I'm rewriting/reorganizing it but there are likely to still be outstanding issues.
<!--- Optional: Include additional context or expand the description here.--->

<!--- After you open your PR, ask for review from the OpenShift docs team:
  For community authors: Tag @openshift/team-documentation in a GitHub comment.--->

apinnick commented on virt/vm_templates/virt-automatic-bootsource-updates.adoc:
```diff
@@ -0,0 +1,31 @@
+
+[id="feature-gate_virt-automatic-bootsource-updates"]
+== Using the feature gate
+
```
Looks much better, Pan! 

I have been giving some thought to the organization of this assembly. I think that organizing the contents as "custom vs system" might not be as clear as I had hoped. I wonder whether this would be clearer:

== Enabling boot source updates

=== Enabling updates for system boot sources by using the feature gate (currently "Toggling automatic updates for system-defined boot sources")

=== Enabling updates for custom boot sources by editing the HCO resource

(Prerequisite: Configuring storage class for custom boot sources, so that would have to be worked in somehow)

== Disabling boot source updates (currently: Disabling an automatic update for a system-defined or user-defined boot source)

kalexand-rh commented:
The `branch/enterprise-4.14` label has been added to this PR.

This is because your PR targets the `main` branch and is labeled for enterprise-4.13. And any PR going into `main` must also target the latest version branch (enterprise-4.14).

If the update in your PR does NOT apply to version 4.14 onward, please re-target this PR to go directly into the appropriate version branch or branches (enterprise-4.x) instead of main.

ousleyp commented:
@aglitke , How is this looking so far? Thanks!

ousleyp commented:
Questions for reviewers:

- Are both patch commands necessary for changing the default storage class?
- Is it "imports and updates" or just updates? What is being imported? is it ok that I removed most instances of "imports and updates" and replaced it with just "updates"?
- Is the `oc get` command accurate in the "Verifying..." module?
- Does the "Verifying" module contradict the "you must add the common template to the HCO CR if you want to disable it" guidance? Does my use of "defined" cover it?

cc @aglitke and @kgoldbla :) thanks!

arnongilboa commented on modules/virt-managing-auto-update-all-system-boot-sources.adoc:
```diff
@@ -0,0 +1,39 @@
+
+Disabling automatic boot source imports and updates can be helpful to reduce the number of logs in disconnected environments or to reduce resource usage.
+
+To disable automatic updates for all system-defined boot sources, turn off the `enableCommonBootImageImport` feature gate by setting the value to `false`. Setting this value to `true` re-enables the feature gate and turns automatic updates back on.
```
It will just stop the DataImportCron updates. DataSources and golden image PVCs won't be touched.
@nunnatsa ack?

nunnatsa replied:
No. The DataImportCronTemplate will be removed from SSP. I almost 100% sure it will then remove the DataImportCron CRs

nunnatsa replied:
This is a question to CDI or SSP team.

But in general - we do want everything deleted in this case. e.g. in disconnected environment, pulling the images will fail anyway.

arnongilboa replied:
Afaik in disconnected env `DVs` still use their `sourceRef` for `DataSource` referring a local `PVC`.
@akrejcir @dominikholler ?

akrejcir replied:
When `DataImportCronTemplate` is removed from SSP CR, `DataSources` are reverted to their default value. So they will not point to PVC that was created by `DataImportCron`.

I don't know about disconnected env.

ousleyp replied:
@arnongilboa @akrejcir , this is a bit over my head :) can you clarify the impact to the user in this situation? 

The way the introduction is currently phrased makes this seem like a harmless action: "Disabling automatic boot source imports and updates can be helpful to reduce the number of logs in disconnected environments or to reduce resource usage."

If the actual impact of turning it off is disruptive, I want to be really clear about the ways in which the user(s) will experience that disruption. Thanks for your help!

arnongilboa replied:
In `disconnected environments`, users wil get `CDIDataImportCronOutdated` alerts, as `DataImportCron` cannot poll or import the latest disk image versions. Setting `enableCommonBootImageImport` to `false` will eliminate these alerts so I guess that was meant by `reduce the number of logs in disconnected environments`. Afaiu, this is not considered disruptive, as DataImportCron was not operational anyway in this case. As @akrejcir & @nunnatsa commented, this will remove `DataImportCronTemplate` from SSP CR, and `DataSources` are reverted to their default value.

ousleyp replied:
> ...this will remove DataImportCronTemplate from SSP CR, and DataSources are reverted to their default value.

Unless there is a clear impact on the user/admin, I don't know if there is a strong case for adding these details to the downstream docs. We tend to avoid documenting what is happening behind the scenes unless it's necessary. Thoughts? @aglitke @arnongilboa @kgoldbla 

I did just add an update to clarify the part about disconnected environments, though.

aglitke replied:
I just tested what happens if you disable the `enableCommonBootImageImport` feature gate.  All DataImportCron objects are deleted.  The PVCs in the openshift-virtualization-os-images namespace are not removed but the DataSource objects are reconciled to point to nonexisting PVCs.  For example, the rhel8 DataSource will be changed so that it points to a non-existent PVC named 'rhel8' whereas it previously pointed to rhel8-06c76d29d511.

I think this is the behavior we want since the admin can optionally  delete the previously imported os images if they would like.  Boot sources can be manually added by creating the appropriate DV to provide the populated PVC with the base name (eg. rhel8).

aglitke replied:
Maybe "DataSources are reset to PVCs which can be manually populated with operating system images."
and "Optionally, after disabling the feature gate, administrators may delete the previously created PVCs"

arnongilboa commented:
> Questions for reviewers:
> 
> * Are both patch commands necessary for changing the default storage class?
Yes, patching the new default class annotation won't remove the old default storage class annotation

> * Is it "imports and updates" or just updates? What is being imported? is it ok that I removed most instances of "imports and updates" and replaced it with just "updates"?
"import" means the first import of a specific source image to a PVC in the golden namespace (openshift-virtualization-os-images), and "update" means the following imports of that source. You may drop some of the "imports" but at least leave some of them:)

> * Is the `oc get` command accurate in the "Verifying..." module?
You need to add "-o yaml":
oc get hco -n openshift-cnv kubevirt-hyperconverged -o yaml

> * Does the "Verifying" module contradict the "you must add the common template to the HCO CR if you want to disable it" guidance? Does my use of "defined" cover it?
> 
@nunnatsa HCO ^^^ 

> cc @aglitke and @kgoldbla :) thanks!

nunnatsa commented:
> > * Does the "Verifying" module contradict the "you must add the common template to the HCO CR if you want to disable it" guidance? Does my use of "defined" cover it?
> 
> @nunnatsa HCO ^^^
> 

I lost you

nunnatsa commented:
^^^ @ousleyp

ousleyp commented:
> > > * Does the "Verifying" module contradict the "you must add the common template to the HCO CR if you want to disable it" guidance? Does my use of "defined" cover it?
> > 
> > 
> > @nunnatsa HCO ^^^
> 
> I lost you

I apologize for phrasing that in a confusing way! I pinged you on gchat but am pasting my message here as well, in case anyone else can help:

The first relevant context is from step 2 of this procedure [[1]](https://ousleyp.github.io/previews/cnv-25648/virt/vm_templates/virt-automatic-bootsource-updates.html#disable-auto-updates-specific-boot-source_virt-automatic-bootsource-updates). It states that if you need to disable automatic updates for a system-defined boot source, you have to add it to the HCO CR before you can disable the updates. Then we have a procedure [[2]](https://ousleyp.github.io/previews/cnv-25648/virt/vm_templates/virt-automatic-bootsource-updates.html#virt-verify-status-bootsource-update_virt-automatic-bootsource-updates) for verifying if a boot source is system-defined or custom that simply involves looking at the HCO CR entry and checking the status.dataImportCronTemplates.status field for that boot source.

My question was if those two things contradict each other; can someone look at the HCO CR to see if something is a common template if they haven't manually added it to the HCO CR? I guess a missing entry might also be another way to determine that it's a common template, though. That just occurred to me. :) I'm realizing now that using "defined" in the title ("Verifying the status of a defined boot source") probably doesn't help, because the common templates are defined on some level, just not necessarily in the HCO CR...

Thanks for your help!

[1] https://ousleyp.github.io/previews/cnv-25648/virt/vm_templates/virt-automatic-bootsource-updates.html#disable-auto-updates-specific-boot-source_virt-automatic-bootsource-updates
[2] https://ousleyp.github.io/previews/cnv-25648/virt/vm_templates/virt-automatic-bootsource-updates.html#virt-verify-status-bootsource-update_virt-automatic-bootsource-updates

ousleyp commented on modules/virt-configuring-storage-class-bootsource-update.adoc:
```diff
@@ -5,13 +5,20 @@
 .Procedure
 
-. Define a new `storageClassName` by editing the `HyperConverged` custom resource (CR).
+. Open the `HyperConverged` custom resource (CR) by running the following command:
```
I've been thinking I should create a snippet for this whole command; that just moved up my to-do list a few slots. 😀 But for now, I've changed one to match the other.

jeana-redhat commented on modules/virt-disable-auto-updates-single-boot-source.adoc:
```diff
@@ -0,0 +1,53 @@
+
+* To disable automatic updates for a custom boot source, remove the boot source from the `spec.dataImportCronTemplates` field. Automatic updates are disabled for custom boot sources by default.
+
+* To disable automatic updates for a system-defined boot source:
```
Is there a way to avoid having the sentence completion dangle into the sub-steps?

ousleyp replied:
Great point; I'm trying this (feedback welcome):
```
. Disable automatic updates for an individual boot source by editing the `spec.dataImportCronTemplates` field.
+
Custom boot source::
* Single step

+
System-defined boot source::
.. Step 1
.. Step 2
```

jeana-redhat replied:
Seems good to me!

jeana-redhat commented on modules/virt-disable-auto-updates-single-boot-source.adoc:
```diff
@@ -0,0 +1,53 @@
+====
+Automatic updates are enabled by default for system-defined boot sources, but these boot sources are not listed in the CR unless you add them.
+====
+.. Set the value of the `dataimportcrontemplate.kubevirt.io/enable` annotation to `false`.
+
++
```
This block is not attached to the previous step. I _think_ you can fix it by removing this line break:
```suggestion
.. Set the value of the `dataimportcrontemplate.kubevirt.io/enable` annotation to `false`.
+
```
I know these are a bit weird though - if it's misbehaving check this bit of [adoc reference](https://docs.asciidoctor.org/asciidoc/latest/lists/continuation/#enclose-in-open-block).

ousleyp replied:
No idea how that line break got in there! Great catch.

jeana-redhat commented on modules/virt-verify-status-bootsource-update.adoc:
```diff
@@ -72,10 +73,12 @@ status: <1>
-<3> Indicates a system-defined boot source.
-<4> Indicates a user-defined boot source.
+<1> Indicates a system-defined boot source.
+<2> Indicates a user-defined boot source.
```
Custom?

jeana-redhat reviewed (commented):
Really nice set of enhancements! Just a few small things to consider from me :)

/remove-label peer-review-in-progress
/remove-label peer-review-needed
/label peer-review-done

ousleyp commented on modules/virt-managing-auto-update-all-system-boot-sources.adoc:
```diff
@@ -0,0 +1,39 @@
+
+.Procedure
+
+* Toggle the feature gate for automatic boot source updates by editing the `HyperConverged` custom resource.
```
Trying not to facepalm too hard at this one 😹 Thank you for pointing that out!

aglitke commented:
lgtm

ousleyp commented on virt/vm_templates/virt-automatic-bootsource-updates.adoc:
```diff
@@ -0,0 +1,43 @@
+By default, {VirtProductName} automatically updates system-defined boot sources. You can opt out of automatic updates for all system boot sources by disabling the `enableCommonBootImageImport` feature gate.
+
+Custom boot sources are not controlled by the feature gate. You must manage them individually by editing the `HyperConverged` custom resource (CR). You can also use this method to manage individual system-defined boot sources.
+
```
Thanks, @apinnick - this makes sense. I decided to create a concept module about automatic boot source updates because the assembly text was getting lengthy. I tried to incorporate your suggestion, though I didn't want to list the boot sources that RH provides in case it changes (also, I think in OKD the RHEL attribute might render as Fedora).

apinnick replied:
In theory, the auto-updated boot sources should not appear in OKD docs because we do not provide this service upstream. We should probably check at some point to see exactly what is being published in OKD Virt.

kgoldbla reviewed (commented):
lgtm

ousleyp commented:
/cherrypick enterprise-4.14

ousleyp commented:
/cherrypick enterprise-4.13

ousleyp commented:
/cherrypick enterprise-4.12
